### PR TITLE
Well refactoring multi segment

### PR DESCRIPTION
**Pull request by GitPaean**
The purpose of the PR is to introduce the class `MultisegmentWells` to handle the multi-segment wells related functions. 

Several major functions, like `updateWellState`, `computeWellFlux`, `computeSegmentFluidProperties`, 
`computeSegmentPressureDelta`, `addWellFluxEq`, `addWellControlEq`, `updateWellControls` has been moved from `BlackoilMultiSegmentModel` to `MultisegmentWells`. 

Further refactoring of the interfaces of the `Wells` classes is still ongoing, while this is a good point to create PR for `MultisegmentWells`. Only the `flow_multisegment` should be affected and it is tested and verified.

### CODE REVIEW

File: CMakeLists_files.cmake [hunk 1]
```diff
@@ -193,6 +194,7 @@ list (APPEND PUBLIC_HEADER_FILES
   opm/autodiff/VFPInjProperties.hpp
   opm/autodiff/WellStateMultiSegment.hpp
   opm/autodiff/WellMultiSegment.hpp
+	opm/autodiff/MultisegmentWells.hpp
```
Comment (atgeirr) [inline CMakeLists_files.cmake]: The _impl.hpp file must be added as well.

File: opm/autodiff/MultisegmentWells.hpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            const std::vector<WellMultiSegmentConstPtr>& wells() const;
+            const MultisegmentWellOps& wellOps() const;
+
+            int numPhases() const { return num_phases_; };
+
+            int numSegment() const { return nseg_total_; };
+
+            int numPerf() const { return nperf_total_; };
+
+            const Vector& wellPerforationCellPressureDiffs() const { return well_perforation_cell_pressure_diffs_; };
```
Comment (atgeirr) [inline opm/autodiff/MultisegmentWells.hpp]: Is it necessary to provide write access to all of these members? I see that some are written to from the model class, but not all of them?

Comment (GitPaean) [inline opm/autodiff/MultisegmentWells.hpp]: Most of them are not, and even should be `protected`.  Some are just needed when they are still in the model class. I will fix them.

**Comment by joakim-hove** [comment]
This is brand new code,  in a complex and important part of OPM: Why oh why are there no tests?

Adding them afterwards is already to late - then the tests can no longer guide you in the design phase.

**Comment by atgeirr** [comment]
Good point @joakim-hove, actually this work is taking the multisegment well stuff from impossible-to-unit-test to possible-to-unit-test. So I guess we should take the opportunity to add tests (other than integration testing), now that we can. We will discuss this next week.

**Comment by GitPaean** [comment]
Added a test for the `MultisegmentWells` class and the previous comments are fixed.

Please suggest. Thanks.

**Comment by atgeirr** [comment]
Thanks for adding a test, even though it is does not test very deeply it is a start, and a skeleton on which to build more tests for this.

I'll merge this PR now.

**Pull request merged**